src/zkstar.rs

```diff
@@ -0,0 +1,118 @@
+use crate::errors::InternalError;
+use libpaillier::unknown_order::BigNumber;
+use serde::{Deserialize, Serialize};
+
+/// A warpper around the modulus of the group [ZStarN] is defined over
+/// This allows passing the modulus by reference and avoid serializing it
+#[derive(Serialize)]
+struct ZStarNBuilder {
+    N: BigNumber,
+}
+
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { N: n }
+    }
+    /// The only way to construct an element of [ZStarN] is to validate that [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.value() {
+            return Err(InternalError::LargerThanModulus);
+        }
+        let result = unverified.value().extended_gcd(self.value());
+        if result.gcd != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().clone(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn value(&self) -> &BigNumber {
+        &self.N
+    }
+}
+
+/// An verifier element of [ZStarN]
+#[derive(Serialize)]
+struct ZStarN<'a> {
+    value: BigNumber,
+    builder: &'a ZStarNBuilder,
+}
+
+impl<'a> ZStarN<'a> {
+    /// Create new canonically-formed element in `ZStar`.
+    /// Constructor always verifies that the element is a member of ZStar
+    fn new(
```

Comment:
For the time being I was only using it in the tests. But I can totally remove it.

Reply:
I'm fine with either removing it or tagging with `#[cfg(test)]`, but I would lean slightly toward removing because this doesn't seem particularly easier to call than the builder version. It's also kind of nice for the test to reflect expected external usage (although that's often not reasonable).